### PR TITLE
xassist implementation

### PR DESCRIPTION
**Pull request by tharun571**
# Description

- Introduced the "xassist" magic command to interact with large language models (LLMs) such as OpenAI and Gemini.
- Integrated the curl library to facilitate HTTP requests to the LLM APIs.
- Implemented functionality to save and load API keys for different models.
- Added support for maintaining context and chat history within the "xassist" command.
- Included necessary documentation to guide users on how to use the new features.
- Added tests to ensure the functionality works as expected and to help users verify their setup.

## Type of change

Please tick all options which are relevant.

- [ ] Bug fix
- [x] New feature
- [x] Added/removed dependencies
- [ ] Required documentation updates

### CODE REVIEW

**Comment by SylvainCorlay** [comment]
I think this shows we need an extension system for magics, so that we don't keep to much in core xeus-cpp and have such features in extensions.

**Comment by vgvassilev** [comment]
Perhaps we need to implement https://github.com/compiler-research/xeus-cpp/issues/4. @tharun571, can you take a look?

I guess here the question is should we accept the PR and move it as a plugin or wait for the plugin system?

**Comment by JohanMabille** [comment]
I think we can accept the PR and move it to the plugin system when we have it, because the implementation does not spread across the whole kernel. My only concern is the dependency on curl, but we can probably hide it behind a building flag if we package a new version of xeus-cpp before the plugin system is available.

**Review by JohanMabille — APPROVED** [review]

File: CMakeLists.txt [hunk 1]
```diff
@@ -203,6 +203,8 @@ set(XEUS_CPP_SRC
     src/xoptions.cpp
     src/xparser.cpp
     src/xutils.cpp
+    src/xmagics/xassist.hpp
```
Comment (vgvassilev) [inline CMakeLists.txt:206]: We should make these conditional to #ifdef __EMSCRIPTEN__ in cmake. Also
We should split the header file and the source file in the previous list.

**Review by vgvassilev — APPROVED** [review]
LGTM, the failures are actually regression in the master (see #155)

**Comment by anutosh491** [comment]
I guess this would involve having a wasm build for curl (to support the esmcripten CI) ? 
I'm giving this an attempt today. Let me try to host a recipe for the same on emscripten-forge if we can get it to work that way.

**Comment by anutosh491** [comment]
Raised a recipe here (https://github.com/emscripten-forge/recipes/pull/1332)

I guess having `libcurl-static` would suffice. We just need the `curl.h` file is it ?

**Comment by tharun571** [comment]
> I guess having `libcurl-static` would suffice. We just need the `curl.h` file is it ?
Yes `curl.h` would help.

**Comment by anutosh491** [comment]
I've added the recipe. Just needs to be reviewed by the folks there but in the meantime, any reviews from here (as to if we generate everything required or not would be great).

Currently I generate 
```
 │ │ Files in package:
 │ │   - bin/curl
 │ │   - bin/curl-config
 │ │   - include/curl/curl.h
 │ │   - include/curl/curlver.h
 │ │   - include/curl/easy.h
 │ │   - include/curl/header.h
 │ │   - include/curl/mprintf.h
 │ │   - include/curl/multi.h
 │ │   - include/curl/options.h
 │ │   - include/curl/stdcheaders.h
 │ │   - include/curl/system.h
 │ │   - include/curl/typecheck-gcc.h
 │ │   - include/curl/urlapi.h
 │ │   - include/curl/websockets.h
 │ │   - lib/libcurl.a
 │ │   - lib/pkgconfig/libcurl.pc
 │ │   - info/about.json
 │ │   - info/hash_input.json
 │ │   - info/index.json
 │ │   - info/licenses/COPYING
 │ │   - info/paths.json
 │ │   - info/recipe/build.sh
 │ │   - info/recipe/recipe.yaml
 │ │   - info/recipe/rendered_recipe.yaml
 │ │   - info/recipe/variant_config.yaml
 │ │   - info/tests/tests.yaml

```

**Pull request merged**